### PR TITLE
Add Rename UI

### PR DESCRIPTION
**Pull request by Colengms**

### CODE REVIEW

File: Extension/src/LanguageServer/references.ts [hunk 1]
```diff
@@ -334,22 +349,38 @@ export class ProgressHandler {
...
-            // Show confirmed references if: user previews results OR peek references is canceled
-            let showConfirmedReferences: boolean =
-                (!this.referencesCanceled && this.referencesCurrentProgress.referencesProgress !== ReferencesProgress.FinalResultsAvailable)
-                || (this.referencesCanceled && this.referencesCurrentProgress.referencesProgress === ReferencesProgress.CanceledFinalResultsAvailable);
-            let peekReferencesResults: string = this.findAllRefsView.getResultsAsText(showConfirmedReferences);
-            if (peekReferencesResults) {
-                this.referencesChannel.appendLine(peekReferencesResults);
+        if (this.client.ReferencesCommandMode === ReferencesCommandMode.Rename) {
+            if (!this.referencesCanceled) {
                 this.referencesChannel.show(true);
```
Comment (michelleangela) [inline Extension/src/LanguageServer/references.ts]: What gets displayed in references channel for rename?

File: Extension/src/LanguageServer/references.ts [hunk 2]
```diff
@@ -325,6 +338,8 @@ export class ProgressHandler {
         this.referencesViewFindPending = false;
         this.referencesChannel.clear();
```
Comment (michelleangela) [inline Extension/src/LanguageServer/references.ts]: Put these in a function that clears all views?

        this.referencesChannel.clear();
        this.findAllRefsView.show(false);
        this.renameView.show(false);

File: Extension/src/LanguageServer/client.ts [hunk 3]
```diff
@@ -574,6 +598,128 @@ export class DefaultClient implements Client {
...
+                                        }
+                                    }
+                                });
+                            });
+                        }
+                    }
+
+                    this.disposables.push(vscode.languages.registerReferenceProvider(documentSelector, new FindAllReferencesProvider(this)));
+
+                    class RenameProvider implements vscode.RenameProvider {
```
Comment (michelleangela) [inline Extension/src/LanguageServer/client.ts:665]: Put RenameProvider in different file?

File: Extension/src/LanguageServer/client.ts [hunk 4]
```diff
@@ -574,6 +598,128 @@ export class DefaultClient implements Client {
                         }
                     }
                     this.disposables.push(vscode.languages.registerWorkspaceSymbolProvider(new WorkspaceSymbolProvider(this)));
+
+                    class FindAllReferencesProvider implements vscode.ReferenceProvider {
```
Comment (michelleangela) [inline Extension/src/LanguageServer/client.ts:602]: Put FindAllReferencesProvider  in different file?

File: Extension/src/LanguageServer/references.ts [hunk 5]
```diff
@@ -109,6 +111,7 @@ export class ProgressHandler {
```
Comment (michelleangela) [inline Extension/src/LanguageServer/references.ts:123]: Let's rename "ProgressHandler" to "ReferencesManager"

File: Extension/src/LanguageServer/renameDataProvider.ts [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+            result.description = true;
+            result.contextValue = "candidateFile";
+            return result;
+        }
+
+        if (element instanceof RenameCandidateReferenceTypeGroupItem) {
+            let label: string = localize("candidates.for.rename", "Candidates for Rename");
+            const result: vscode.TreeItem = new vscode.TreeItem(label);
+            result.collapsibleState = vscode.TreeItemCollapsibleState.Expanded;
+            result.iconPath = null;
```
Comment (michelleangela) [inline Extension/src/LanguageServer/renameDataProvider.ts]: you can delete result.iconPath if it's null

File: Extension/src/LanguageServer/renameDataProvider.ts [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+            return result;
+        }
+
+        if (element instanceof RenamePendingFileItem) {
+            const result: vscode.TreeItem = new vscode.TreeItem(element.uri);
+            result.collapsibleState = vscode.TreeItemCollapsibleState.Expanded;
+            result.iconPath = vscode.ThemeIcon.File;
+            result.description = true;
+            result.contextValue = "pendingFile";
+            let tag: string = getReferenceTagString(ReferenceType.ConfirmationInProgress, false);
```
Comment (michelleangela) [inline Extension/src/LanguageServer/renameDataProvider.ts]: Pass in element.type instead of "ReferenceType.ConfirmationInProgress"?

**Review by michelleangela — CHANGES_REQUESTED** [review]

File: Extension/src/LanguageServer/client.ts (hunk 3 above)
Comment (Colengms) [reply Extension/src/LanguageServer/client.ts:665]: All of these providers need to be in their own files.  I'd like to do that in a subsequent PR, after Insiders.

File: Extension/src/LanguageServer/client.ts (hunk 4 above)
Comment (Colengms) [reply Extension/src/LanguageServer/client.ts:602]: Same as above.  The other 3 as well.  But, there are some complications I'd like to address after Insiders.

File: Extension/src/LanguageServer/renameDataProvider.ts (hunk 7 above)
Comment (Colengms) [reply Extension/src/LanguageServer/renameDataProvider.ts]: This is actually the file level.  I removed the tooltip for it.

**Review by sean-mcmanus — APPROVED** [review]

**Pull request merged**